### PR TITLE
Run connector builder server tests and static analysis from a docker container

### PR DESCRIPTION
**Pull request by girarda**
## What
Run the connector-builder-server's tests, code formatting, and static analysis from a docker container to remove the python dependency from the platform build toolchain.

## How
* Create a new gradle plugin `airbyte-python-docker` which runs the same commands as the `airbyte-plugin` inside a docker container with Python pre-installed
* The `airbyte-python-docker` has a single step which runs a script from a docker container
* The script runs through the same commands that are in [airbyte-python](https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle) (minus the static analysis reports)

## Recommended reading order
1. `buildSrc/src/main/groovy/airbyte-python-docker.gradle`
2. `airbyte-connector-builder-server/run_tests.sh`
3. `airbyte-connector-builder-server/build.gradle`
4. `airbyte-connector-builder-server/pyproject.toml`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: buildSrc/src/main/groovy/airbyte-python-docker.gradle [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+
+        // attempt to install anything in requirements.txt. by convention this should only be dependencies whose source is located in the project.
+
+        if (project.file('requirements.txt').exists()) {
+            throw new GradleException("Unexpected")
+        } else if (project.file('setup.py').exists()) {
+            project.task("installLocalReqs", type: Exec) {
+                // FIXME: use pwd to get the path
+                commandLine "docker", "run", "-v", "/Users/alex/code/airbyte/airbyte-connector-builder-server:/tmp", "--entrypoint", "pip", "python:3.9-slim", "install", "-e", "/tmp"
```
Comment (girarda) [inline buildSrc/src/main/groovy/airbyte-python-docker.gradle:152]: @sherifnada this is how we can run the python command from inside a docker container

File: airbyte-connector-builder-server/run.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+pip install 'flake8==4.0.1'
+pip install 'pyproject-flake8==0.0.1a2'
+pip install 'black==22.3.0'
+pip install 'mypy==0.930'
+pip install 'isort==5.6.4'
+pip install 'pytest==6.1.2'
+pip install 'coverage[toml]==6.3.1'
+apt update
+apt install -y git
+python -m isort --settings-file=pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run.sh]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L100-L103

File: airbyte-connector-builder-server/run.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+python -m isort --settings-file=pyproject.toml --diff --quiet ./
+python -m black --config pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run.sh]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L111-L117

File: airbyte-connector-builder-server/run.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+python -m black --config pyproject.toml ./ --diff --quiet
```
Comment (girarda) [inline airbyte-connector-builder-server/run.sh]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L119-L123

File: airbyte-connector-builder-server/run.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+python -m pflake8 --config pyproject.toml ./
+python -m pflake8 --config pyproject.toml ./ --diff --quiet
```
Comment (girarda) [inline airbyte-connector-builder-server/run.sh]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L130-L134

File: airbyte-connector-builder-server/run_tests.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+pip install 'black==22.3.0'
+pip install 'mypy==0.930'
+pip install 'isort==5.6.4'
+pip install 'pytest==6.1.2'
+pip install 'coverage[toml]==6.3.1'
+
+# Format and static analysis
+python -m isort --settings-file=pyproject.toml ./
+python -m isort --settings-file=pyproject.toml --diff --quiet ./
+python -m black --config pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:25]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L111-L117

File: airbyte-connector-builder-server/run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+python -m black --config pyproject.toml ./ --diff --quiet
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:26]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L119-L123

File: airbyte-connector-builder-server/run_tests.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+pip install 'pyproject-flake8==0.0.1a2'
+pip install 'black==22.3.0'
+pip install 'mypy==0.930'
+pip install 'isort==5.6.4'
+pip install 'pytest==6.1.2'
+pip install 'coverage[toml]==6.3.1'
+
+# Format and static analysis
+python -m isort --settings-file=pyproject.toml ./
+python -m isort --settings-file=pyproject.toml --diff --quiet ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:24]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L105-L109

File: airbyte-connector-builder-server/run_tests.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+pip install 'flake8==4.0.1'
+pip install 'pyproject-flake8==0.0.1a2'
+pip install 'black==22.3.0'
+pip install 'mypy==0.930'
+pip install 'isort==5.6.4'
+pip install 'pytest==6.1.2'
+pip install 'coverage[toml]==6.3.1'
+
+# Format and static analysis
+python -m isort --settings-file=pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:23]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L100-L103

File: airbyte-connector-builder-server/run_tests.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+python -m isort --settings-file=pyproject.toml --diff --quiet ./
+python -m black --config pyproject.toml ./
+python -m black --config pyproject.toml ./ --diff --quiet
+python -m pflake8 --config pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:27]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L125-L128

File: airbyte-connector-builder-server/run_tests.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+python -m pflake8 --config pyproject.toml ./ --diff --quiet
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:28]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L130-L134

File: airbyte-connector-builder-server/run_tests.sh [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+python -m mypy --config pyproject.toml ./
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:29]: https://github.com/airbytehq/airbyte/blob/master/buildSrc/src/main/groovy/airbyte-python.gradle#L136-L140

File: airbyte-connector-builder-server/connector_builder/impl/default_api.py [hunk 13]
```diff
@@ -9,19 +9,24 @@
 from urllib.parse import parse_qs, urljoin, urlparse
 
 from airbyte_cdk.models import AirbyteLogMessage, AirbyteMessage, Type
+from fastapi import Body, HTTPException
```
Comment (girarda) [inline airbyte-connector-builder-server/connector_builder/impl/default_api.py:12]: all changes to this file are from gradle fromat

File: airbyte-connector-builder-server/run_tests.sh [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+pip install 'flake8==4.0.1'
+pip install 'pyproject-flake8==0.0.1a2'
+pip install 'black==22.3.0'
+pip install 'mypy==0.930'
+pip install 'isort==5.6.4'
+pip install 'pytest==6.1.2'
+pip install 'coverage[toml]==6.3.1'
+
+# Format and static analysis
+# FIXME: isort formats python files differently from gradlew format
```
Comment (girarda) [inline airbyte-connector-builder-server/run_tests.sh:23]: Still need to figure out why gradle format and isort are using different rules

File: airbyte-connector-builder-server/pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,83 @@
+[tool.black]
```
Comment (girarda) [inline airbyte-connector-builder-server/pyproject.toml:1]: this file is a copy of https://github.com/airbytehq/airbyte/blob/master/pyproject.toml without the [skip_gitignore to remove the dependency on git](https://github.com/airbytehq/airbyte/blob/master/pyproject.toml#L44)

File: buildSrc/src/main/groovy/airbyte-python-docker.gradle [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.Exec
+
+class AirbytePythonDockerConfiguration {
```
Comment (sherifnada) [inline buildSrc/src/main/groovy/airbyte-python-docker.gradle:5]:
```suggestion
/*
This plugin exists to build & test a python module via Gradle without requiring the developer to install Python on their machine. It achieves this by mounting the module into a Python Docker image and running the tests. 

Modules consuming this plugin must define two scripts `run_tests.sh` and `run_format.sh`. 
*/

class AirbytePythonDockerConfiguration {
```

File: airbyte-connector-builder-server/run_tests.sh [hunk 17]
```diff
@@ -0,0 +1,10 @@
+cd $1
+
+# Install dependencies
+pip install -e .
+pip install -e '.[main]'
+pip install -e '.[tests]'
+
+# Run the tests
+python -m coverage run -m pytest unit_tests -c pytest.ini
+python -m coverage run -m pytest integration_tests -c pytest.ini
```
Comment (sherifnada) [inline airbyte-connector-builder-server/run_tests.sh:10]:
```suggestion
python -m coverage run -m pytest integration_tests -c pytest.ini

```

File: airbyte-connector-builder-server/run_format.sh [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+# Format and static analysis
+# FIXME: isort formats python files differently from gradlew format
+python -m isort --settings-file=pyproject.toml ./
+python -m isort --settings-file=pyproject.toml --diff --quiet ./
+python -m black --config pyproject.toml ./
+python -m black --config pyproject.toml ./ --diff --quiet
+python -m pflake8 --config pyproject.toml ./
+python -m pflake8 --config pyproject.toml ./ --diff --quiet
+python -m mypy --config pyproject.toml ./
```
Comment (sherifnada) [inline airbyte-connector-builder-server/run_format.sh:21]:
```suggestion
python -m mypy --config pyproject.toml ./

```

File: airbyte-connector-builder-server/build.gradle [hunk 19]
```diff
@@ -42,4 +42,18 @@ task generateOpenApiPythonServer(type: GenerateTask){
 }
 
 project.build.dependsOn(generateOpenApiPythonServer)
-    
+
+task prepareBuild(type: Copy) {
```
Comment (sherifnada) [inline airbyte-connector-builder-server/build.gradle:46]: why is this needed? can you add a comment with the context?

**Review by sherifnada — APPROVED** [review]

File: airbyte-connector-builder-server/build.gradle (hunk 19 above)
Comment (lmossman) [reply airbyte-connector-builder-server/build.gradle:50]: I left a comment [here](https://github.com/airbytehq/airbyte/pull/19309#pullrequestreview-1176626476) asking if this step meant that if in the future we add some new file or directory that we need to be in the docker image, then we will need to update this step as well. And you said that you could replace this with just a `"."` to copy everything in the directory over.

Just curious, is there a reason you didn't go with that approach here? (Also, is that other PR being replaced with this one?)

File: airbyte-connector-builder-server/integration_tests/test_integration_test.py [hunk 20]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+
+def test_test():
```
[the remark this answers is not in the document]
Comment (lmossman) [reply airbyte-connector-builder-server/integration_tests/test_integration_test.py:6]: (I heard you like tests, so I made a ...^)

File: airbyte-connector-builder-server/pyproject.toml (hunk 15 above)
Comment (lmossman) [reply airbyte-connector-builder-server/pyproject.toml:1]: Will it be annoying to maintain these two different files separately since they have so much duplication? I'm guessing there isn't an easy way to configure this to just ignore that one line?

Also totally fine if the answer is just "this is the quickest way to get V0 working"

**Review by lmossman — COMMENTED** [review]
Had a few very minor comments

File: airbyte-connector-builder-server/build.gradle (hunk 19 above)
Comment (girarda) [reply airbyte-connector-builder-server/build.gradle:50]: copy "." copies a bunch unwanted hidden files. Will look for an exclude filter because i still agree with you

File: airbyte-connector-builder-server/pyproject.toml (hunk 15 above)
Comment (girarda) [reply airbyte-connector-builder-server/pyproject.toml:1]: fair point. The reason I copied it is that I removed the dependency on git. Thinking more about it, I think it'll be easier to maintain if we just install git in the docker container and symlink to the original

File: airbyte-connector-builder-server/build.gradle (hunk 19 above)
Comment (girarda) [reply airbyte-connector-builder-server/build.gradle:49]: added an exclude filter
